Trash Manager (restored from #13515)

wuuti opened the pull request:
Restored original work in a new branch, original PR was https://github.com/modxcms/revolution/pull/13515.

### What does it do?
Adds a _Trash_ menu entry and a Trash managing page. The page lists all deleted resources together with their contexts, pagetitle, longtitle, the deletion date and the username who deleted the resource.

<img width="1001" alt="screen shot 2017-07-07 at 15 33 56" src="https://user-images.githubusercontent.com/2852461/27960001-ea586064-6329-11e7-8169-472d22d269c8.png">
<img width="759" alt="screen shot 2017-07-07 at 15 34 22" src="https://user-images.githubusercontent.com/2852461/27960003-eda978e8-6329-11e7-9a11-23243b037381.png">
<img width="757" alt="screen shot 2017-07-07 at 15 34 43" src="https://user-images.githubusercontent.com/2852461/27960005-ef68c68e-6329-11e7-9054-fad259fc7787.png">

Resources can be restored or purged by the context menu, by the bulk actions or all resources can be purged with a button. The button fullfils the same functionality as the bin icon above the resource tree in previous versions. 
With this merge the behaviour of the bin icon changes, it now directs you to the trash manager page. 

### Why is it needed?
The former behaviour was an all or nothing purge. It was not possible, for example, to purge only resources from a certain context. This can now be easily done. Also the restoring of resources was not possible for multiple resources at once. And if deleted resources where hidden from the tree (e.g. inside a collections container), users may completely overlook them. 

### Related issue(s)/PR(s)
Solves parts of the multi-issue #764.
#13617, #13491

Commit:
Restored from broken branch.

Commit:
Added missing file.

Commit:
Removed wrong file.

Commit:
Added sort to access policy.

Commit:
Merge remote-tracking branch 'origin/trash-restored' into trash-restored

# Conflicts:
#	manager/assets/modext/widgets/resource/modx.tree.resource.js

Commit:
Removed wrong tempfile.

Mark-H commented on core/model/modx/processors/resource/trash/getlist.class.php:
```diff
@@ -0,0 +1,139 @@
+        $cls[] = 'trashrow';
+
+
+        $debug = array('id'=>$object->get('id'), 'pagetitle'=>$object->get('pagetitle'));
```
These debugs should be removed

Mark-H commented on core/model/modx/processors/resource/trash/purge.class.php:
```diff
@@ -0,0 +1,210 @@
+            return $this->modx->lexicon('resource_err_ns');
+        }
+        if ($idlist==-1) {
+            $this->modx->log(modX::LOG_LEVEL_WARN, "[purge] purging everything we are allowed to.");
```
This (and a couple of other lines below) will send data to the MODX Error log (if WARN is enabled) - would be better to log to the manager log instead.

Mark-H commented on core/model/modx/processors/resource/trash/purge.class.php:
```diff
@@ -0,0 +1,210 @@
+            // TODO isn't that a problem here?
+            // If resource remove now fails we already removed the tvs!
+            // shouldn't resource->remove also take care of the tvs and resource groups?
+            if ($resource->remove()==false) {
```
Code style: `if (!$resource->remove()) {` is simpler

Mark-H commented on core/model/modx/processors/resource/trash/purge.class.php:
```diff
@@ -0,0 +1,210 @@
+        }
+
+        $this->modx->invokeEvent('OnEmptyTrash', array(
+            'num_deleted' => sizeof($success),
```
`sizeof` is an alias of `count` - so please just use `count`.

Mark-H commented on core/model/modx/processors/resource/trash/purge.class.php:
```diff
@@ -0,0 +1,210 @@
+        }
+
+        $msg = "no message";
+        if (sizeof($success)>0) {
```
`sizeof` > `count` again

Mark-H commented on core/model/modx/processors/resource/trash/purge.class.php:
```diff
@@ -0,0 +1,210 @@
+            return $this->success($this->modx->lexicon('trash.purge_err_nothing'));
+        }
+
+        $msg = "no message";
```
Better to just set it to an empty string if there's no message, as semantically, `no message`, is a message.

Mark-H commented on core/model/modx/processors/resource/trash/restore.class.php:
```diff
@@ -0,0 +1,174 @@
+    {
+        // we expect a list of files here
+        $this->idList = explode(',', $this->getProperty('ids', false));
+        $this->modx->log(1, "Restorelist: " . print_r($this->idList, true));
```
Debug should be removed

Mark-H commented on core/model/modx/processors/resource/trash/restore.class.php:
```diff
@@ -0,0 +1,174 @@
+            'deleted' => true,
+            'id:IN'   => $this->idList,
+        ));
+        $count = sizeof($this->resourceList);
```
`sizeof` > `count`

Mark-H commented on core/model/modx/processors/resource/trash/restore.class.php:
```diff
@@ -0,0 +1,174 @@
+            return $this->modx->lexicon('resource_err_nfs', array('ids' => $this->idList));
+        }
+
+        $this->modx->log(1, "found " . $count);
```
debug should be removed

Mark-H commented on core/model/modx/processors/resource/trash/restore.class.php:
```diff
@@ -0,0 +1,174 @@
+        $this->modx->log(1, "found " . $count);
+
+        foreach ($this->resourceList as $resource) {
+            $this->modx->log(1, "restoring " . $resource->get('id') . ": " . $resource->get('pagetitle'));
```
as in the other place, probably better sent to the manager log instead of error log.

Mark-H commented on core/model/modx/processors/resource/trash/restore.class.php:
```diff
@@ -0,0 +1,174 @@
+        }
+
+        /* validate resource can be deleted */
+        //if (!$this->resource->checkPolicy(array('save' => true, 'delete' => true))) {
```
why is this commented out? If not needed, please remove.

wuuti replied:
Well, that was a try to check for fine grained permissions for that file. But it turns out that permissions are not evaluated as they should here. I will open up another issue for that (as discussed at the snowup 2018)

Mark-H commented on core/model/modx/processors/resource/trash/restore.class.php:
```diff
@@ -0,0 +1,174 @@
+            $this->resource->set('deletedby', 0);
+            $this->resource->set('deletedon', 0);
+
+            if ($this->resource->save()==false) {
```
`if (!$this->resource->save()) {` is simpler

Mark-H commented on manager/assets/modext/sections/resource/trash/index.js:
```diff
@@ -0,0 +1,24 @@
+        components: [{
+            xtype: 'modx-panel-trash'
+        }]
+        /*,buttons: [{
```
If not used, it's cleaner to just remove entirely instead of commenting out.

Mark-H commented on manager/assets/modext/widgets/resource/modx.panel.trash.js:
```diff
@@ -0,0 +1,390 @@
+    config = config || {};
+    Ext.applyIf(config, {
+        id: 'modx-panel-trash'
+        , cls: 'container'
```
Inconsistent code style in this section with the space between the comma and the property

Mark-H commented on manager/assets/modext/widgets/resource/modx.panel.trash.js:
```diff
@@ -0,0 +1,390 @@
+        var trashButton = t.getTopToolbar().findById('emptifier');
+
+        var count = this.getStore().getTotalCount();
+        console.log("Items left: " + total)
```
Stray console.log statement

Mark-H commented on manager/assets/modext/widgets/resource/modx.tree.resource.js:
```diff
@@ -227,7 +227,8 @@ Ext.extend(MODx.tree.Resource,MODx.tree.Tree,{
 
-                        trashButton.setTooltip(_('empty_recycle_bin') + ' (' + data.object.deletedCount + ')');
+
+                        //    trashButton.setTooltip(_('empty_recycle_bin') + ' (' + data.object.deletedCount + ')');
```
remove instead of commenting out if no longer needed

Mark-H reviewed (dismissed):
Looking promising, but have done a code review and would appreciate it if you could look at and correct the comments. Thanks!

Oetzie commented:
Why a special cmp for this, is this not overkill? Was it not easier to just make a popup when you click on the trashcan? In that popup you can see a tree with the resources and checkboxes that you need to confirm?

wuuti commented:
## Current state:
- deleted resources are listed at the trash manager page
- the list respects the context access rights: if you don't have access to a context, deleted resources from that context won't be listed at the trash manager page!
- deleting/undeleting a resource in the tree automatically refreshes the trash manager list (if the trash manager page is open while you delete/undelete)
- restoring/purging an item from the trash manager list refreshes the tree automatically
- trash manager is only accessible with the "purge_deleted" rights

## Wishlist:
- optional handling of child resources on purge or undelete
- automatically refresh the trash list if someone else deletes / restores files (would need some kind of polling for any changes...)

Commit:
Added policy permission to show the trash manager menu.

Commit:
Fixed bugs with single resource restore.
Added lexicon strings.

Commit:
Linked recycle bin icon to trash manager page.

Commit:
Fixed logging messages.

Commit:
Code cleanup.

Commit:
Implemented code review advices.

wuuti commented:
@Mark-H implemented your reviews.

wuuti commented:
It is also possible to directly unpublish or publish a resource from the trash manager. This is important if you want to restore a formerly published resource and first want to review it before it is published again: 
<img width="708" alt="screen shot 2018-02-10 at 16 06 29" src="https://user-images.githubusercontent.com/2852461/36063387-751aa060-0e7c-11e8-8ad6-d0478a66096e.png">

The recycle bin icon now does not execute any action, instead it brings you to the trash manager page.

OptimusCrime commented:
This is so cool. One of the coolest new features I have seen in the manager in a long time. Awesome work. I'll try to give it a spin when I have the time.

bezumkin commented on manager/assets/modext/widgets/resource/modx.panel.trash.js:
```diff
@@ -0,0 +1,442 @@
+            ,url: this.config.url
+            ,params: {
+                action: 'resource/trash/purge'
+                ,id: this.menu.record.id
```
Purge processor expectes `ids` variable, so here will be an error if I`m trying to remove only one resource with context menu
![screen shot 2018-04-06 at 14 02 55](https://user-images.githubusercontent.com/1257284/38420559-6e233b5c-39a4-11e8-9a19-d104042b6c36.png)

Changing `id` to `ids` solves problem.

bezumkin commented on core/model/modx/processors/resource/trash/restore.class.php:
```diff
@@ -0,0 +1,196 @@
+        return true;
+    }
+
+    public function process() {
```
I suggest you to replace it to something like this:
```php
foreach ($ids as $id) {
    $response = $this->modx->runProcessor('resource/undelete', array('id' => $id));
    if ($response->isError()) {
        return $response->getResponse();
    }
}
return $this->modx->success();
```
We already have undelete processor logic, why just don`t use it? It is much simpler and allows to avoid duplicate of logic.

And maybe it is time to add `resource/remove` processor and use it in `resource/emptyrecyclebin` and `resource/trash/purge`

gpsietzema commented:
@wuuti do you possibly have some time to resolve the above feedback? We want to merge everything for 2.7 asap and it would be wonderful to have this in there.

Mark-H commented:
Closing as the work on this has continued in #14058
